### DISCUSSION TITLE
Weird output from VST3 restricted instances

### DISCUSSION THREAD

**Discussion by dervur** [RomanKubiak/ctrlr#675] [Q&A] [answered]
Hello there, 
I don't know if this is a bug or I'm doing something wrong. I was updating one of my panels to include channel selection. While making the video for YouTube I realized that selecting the wrong channel instead of not controlling the device it was still controlling it albeit weirdly...

I downloaded MIDI view and this is what I have at the output of my VST3 restricted instance and...
![VST3](https://github.com/user-attachments/assets/7619c136-0858-4483-b58a-02e75d4b3b7c)

In the VST I had selected channel 5 (while my device is on channel 1), you can see that it randomly controls something on channel 1 at each step of me turning a knob on the VST set to channel 5...

I went to the VST2 restricted instance of the same panel, and the output is instead correct.
![VST2](https://github.com/user-attachments/assets/16f091f8-4c5d-48cf-9446-649f9720190b)

I then opened CtrlrX in VST3, and the panel from there, the behaviour was correct as in the VST 2 restricted instance. It appears that this bizzarre behaviour is triggered only by the VST3 restricted instances... Did anybody experience this before? Is there a solution? It seems very weird that it controls something randomly at a different channel... I also tried on other exported panels, and the behaviour is the same, so it happears to be a VST3 restricted instance only problem... even the exported exe works as designed.

**Comment by damiensellier** [comment]
That's really strange, thanks for pointing out. I'll give it a look when back from vacations. 
Is everything fine with your project apart from that?

**Reply by dervur** [reply]
Hi Damien, yeah there's no rush, enjoy your holidays.
Yeah everything's fine, I might release the update without VST3 since I think channel selection is quite important and should've been there already...

**Comment by damiensellier** [comment]
Are you using LUA to send midi messages?
In this case, maybe you should try switching off the modulators' midi outputs : 

![Capture d’écran, le 2024-08-07 à 23 14 14](https://github.com/user-attachments/assets/e811716f-b53f-4404-abbb-a7a5e2ea310f)

It's not really related but who knows. 
I'll go through all the "if isRestricted()" in xcode and will see if there's something wrong about midi messages.

**Reply by dervur** [reply]
Mmmh no, not using LUA to send MIDI messages, except when I re-channel the input (which actually doesn't work on VSTs anyway so I should just deactivate it on VST instances). But that is not related to any modulators but only incoming MIDI messages.

I will try to mute the combo boxes and buttons that do not send MIDI albeit they are already set to midi message "none", and export a restricted instance with that and see if it helps. But as the problem is not shown on VST2 and exe instances, I think there is still something wrong in the code if modulator set to "non" midi message type send something random at a random channel.

**Reply by damiensellier** [reply]
I remember having duplicate messages when running 2 instances of ctrlr at the same time. One as VST and another one in the background as App. 

I'll check anyway what can be wrong about that.

**Reply by damiensellier** [reply]
@dervur I thought about that yesterday while working on CtrlrX new release.
This problem could also be related to MIDI Thru options doubling the message please check your MIDI Thru Host options that you did not enable anything special : 

![MIDI THRU VST3](https://github.com/user-attachments/assets/cad91833-9df9-43d0-807d-d9bd396bf7c6)

**Reply by dervur** [reply]
I will have a better look but this would not explain why the behaviour does not happen in VST2 restricted instances and does not happen within CtrlrX vst3 controller. Also the signal is not repeated, it sends messages always in channel 1 and a different CC, you can see in the first picture the panel is set to channel 5 and I'm turning a knob controlling CC46, the weird output is in channel 1 and the CC of expression (I believe that is CC11).
It is not repeating the output it is creating something else entirely somehow. But this afternoon I will have a look, I'll export a restricted instance with the bar on top so I can have more control options and see if in the export something changes.

I tried a few months ago to look at the CtrlrX code but I couldn't see the reason behind this behaviour. If you can point me to the right files to look at I can try to have a second look. I will also try with the new version of CtrlrX if this problem still occurs.

**Reply by dervur** [reply]
I tried last evening, even with the new version the issue persists, I tried to export one with a menu bar and there is no weird routing apparently...

**Comment by dervur** [comment]
@damiensellier I think I found the issue! Albeit, not the direct solution.

See the issue I posted yesterday about SysEx messages that were making CtrlrX crash. After a fair amount of crashes I noticed this morning that I had the same issue described here for VST3 on the desktop version of CtrlrX (not even a restricted instance).
After I cleared out the settings folder of Ctrlr from AppData, this issue stopped.

My theory is that after a certain amount of crashes, something breaks in the settings and begins this weird behaviour. I think it happens at random. And because of the "panelIsDirty" issue that we managed to solve last summer, both my VST3 and VST2 crashed a lot of times too, but whatever breaks in the settings it only broke in the VST3 version, not the VST 2.
This would explain why the CtrlrX VST3 version was not broken, but only the restricted instance (as only the latter was crashing).

Now the issue is... Do you know where are the settings for the VST3 version stored? I managed to solve it for the desktop version today because I knew where the settings are for that. But the VST3 remains unsolved.
I saw there are folders with my plugin name on AppData as well, but deleting those did not solve anything. So the settings for VST3 (and VST2 for that matter) should be stored somewhere else. I strongly believe that deleting that folder will solve this issue...

**Reply by dervur** [reply, marked as answer]
Ok, I deleted all the CtrlrX folder, even the ones from other panels. And the issue seems to be resolved for now. 
I seem to have another issue now with the MIDI thru, but I'll try to solve it later, and if I don't manage to I will start another thread.

Thank you very much for all the help. If somebody had this issue in future, just go and delete all the settings of all Ctrlr related folders, even the ones from panels that are not affected. I'm glad this issue was finally resolved and I can now release VST3 version of my panels.

**Reply by damiensellier** [reply]
> Do you know where are the settings for the VST3 version stored?

the settings are stored straight into your DAW project file. if cubase for ex, look in your project .cpr

Open it with a text editor and search something from ctrlr like "panelisdirty". you'll find the ctrlr related data part stored via **stateData** to the host.

The default settings like midi settings can also be stored in your plugin default preset settings.

There's no folder with ctrlr.settings associated to the vst vst3 or AU plugins. it's only for thre standalone app.
